src/netbase.cpp

```diff
@@ -48,36 +48,59 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
     ai_hint.ai_protocol = IPPROTO_TCP;
     // We don't care which address family (IPv4 or IPv6) is returned
     ai_hint.ai_family = AF_UNSPEC;
+
     // If we allow lookups of hostnames, use the AI_ADDRCONFIG flag to only
     // return addresses whose family we have an address configured for.
     //
     // If we don't allow lookups, then use the AI_NUMERICHOST flag for
     // getaddrinfo to only decode numerical network addresses and suppress
     // hostname lookups.
-    ai_hint.ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
+    const int ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
+
+    // getaddrinfo is called twice and to prevent duplicates a set is used
+    // std::set used over a std::unordered_set as CNetAddr does not have a hash function
+    std::set<CNetAddr> resolved_addresses_set;
+
+    auto callgetaddrinfo = [&](const int flags, const bool only_add_local_addr) {
+        ai_hint.ai_flags = flags;
+        addrinfo* ai_res{nullptr};
+        getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res);
+
+        // Traverse the linked list of addrinfo structures
+        addrinfo* ai_trav{ai_res};
+        while (ai_trav != nullptr) {
+            CNetAddr addr;
+            if (ai_trav->ai_family == AF_INET) {
+                assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in));
+                addr = CNetAddr{reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr};
+            }
+            if (ai_trav->ai_family == AF_INET6) {
+                assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in6));
+                const sockaddr_in6* s6{reinterpret_cast<sockaddr_in6*>(ai_trav->ai_addr)};
+                addr = CNetAddr{s6->sin6_addr, s6->sin6_scope_id};
+            }
+            if (!only_add_local_addr || (only_add_local_addr && addr.IsLocal())) {
```

Comment:
nit, feel free to ignore if you find the current one more readable, but this is equivalent to:

```suggestion
            if (!only_add_local_addr || addr.IsLocal()) {
```